### ISSUE TITLE
Can I create multiple copies of a device?

### ISSUE THREAD

**Issue by jazz-soft** [micbuffa/WebAudioPlugins#2]
When I try to create two copies of the same device, it fails on the second loadGUI() call.
Is it possible to allow multiple copies?

**Comment by micbuffa** [comment]
Can you give me details ? If I clone the fork of the repo you've done, can I try to run one of your example that fails?
The problem is certainly due to the fact that you need to reload the whole plugin (do the new again, create another instance of the factory, do the load and loadGUI again), and you did not encapsulate the JJZ dependency. The GUI part is a WebComponent, so everything is isolated there. If in this part you include a lib, and if you load multiple instances, this lib will be loaded only once, etc.
But we wanted to separate the DSP and GUI part, because in some projects, we just reuse the DSP part of plugins (i.e in the guitar amp sim, the tonestack is a plugin, it's just a subgraph in the middle of the global audio graph of the sim).
However, as the DSP part is just a JS file, you need to take care of encapsulation yourself: put everything in a class. If you include the JJZ.js file, and create two instances, you'll need to be sure that you're not including JJZ.js twice, or that this double inclusion will not raise conflicts.
You can include the JJZ.js file in the html file of the host, for factorizing and avoiding this problem.

Let me know what I can try to help...

All our plugins can be loaded more than once, you can try this in the pedalboard app (https://wasabi.i3s.unice.fr/dynamicPedalboard/) or if you change some code in the testers...

Michel

**Comment by jazz-soft** [comment]
I did not commit the failing code, but it looks like this:

    <!DOCTYPE html>
    <html>
    
    <head>
      <title>Host that loads a plugin with its GUI</title>
    </head>
    <script src="../sdk/WebAudioSDK.js"></script>
    <script src="JZZ/JZZ.js"></script>
    <script src="JZZ/JZZ.input.Kbd.js"></script>
    <script src="JZZ/JZZ.synth.OSC.js"></script>
    <script src="MidiKeyboard/main.js"></script>
    <script src="../plugins/PureJS/GuitarTuner/main.js"></script>
    <script src="../plugins/PureJS/PingPongDelay/main.js"></script>
    
    <body>
    </body>
    
    <script>
      var ctx = new AudioContext();
      var piano1 = new JazzSoftMidiKeyboard(ctx, 'MidiKeyboard');
      var piano2 = new JazzSoftMidiKeyboard(ctx, 'MidiKeyboard');
      var tuner = new WasabiTunerMachine(ctx, '../plugins/PureJS/GuitarTuner');
      var delay = new WasabiPingPongDelay(ctx, '../plugins/PureJS/PingPongDelay');
    
      Promise.all([piano1.load(), piano2.load(), tuner.load(), delay.load()]).then((nnn) => {
        piano1.loadGui().then((elem) => { document.body.appendChild(elem); });
        piano2.loadGui().then((elem) => { document.body.appendChild(elem); });
        tuner.loadGui().then((elem) => { document.body.appendChild(elem); });
        delay.loadGui().then((elem) => { document.body.appendChild(elem); });
    //    nnn[1].connect(nnn[2]);
    //    nnn[2].connect(nnn[3]);
    //    nnn[3].connect(ctx.destination);
      });
    </script>
    
    </html>

The rest of the code is in my fork.

And, BTW, what is the best place for questions/answers? - I'll probably have a lot more questions...

Thank you!
sema

**Comment by jazz-soft** [comment]
You don't need to clone my fork, I'm getting the same error with your tutorial example (from which I copied):

    <!DOCTYPE html>
    <html>
    
    <head>
      <title>Host that loads a plugin with its GUI</title>
    </head>
    <script src="../../sdk/WebAudioSDK.js"></script>
    <script src="main.js"></script>
    
    <body>
    <audio src="../../assets/audio/BasketCaseGreendayriffDI.mp3" id="soundSample" controls loop crossorigin></audio>
    </body>
    <script>
    
      var ctx = new AudioContext();
      var pluginURL = ".";
      var plugin1 = new WasabiLowfilter(ctx, pluginURL);
      var plugin2 = new WasabiLowfilter(ctx, pluginURL);
    
      var player = document.getElementById("soundSample");
      player.onplay = () => {
        ctx.resume().then(() => {
          console.log('Playback resumed successfully');
        });
      }
      var mediaSource = ctx.createMediaElementSource(player);
    
      Promise.all([plugin1.load(), plugin2.load()]).then((nnn) => {
        plugin1.loadGui().then((elem) => { document.body.appendChild(elem); });
        plugin2.loadGui().then((elem) => { document.body.appendChild(elem); });
        mediaSource.connect(nnn[0]);
        nnn[0].connect(nnn[1])
        nnn[1].connect(ctx.destination)
      });
    </script>
    </html>

Thanks,
sema

**Comment by micbuffa** [comment]
Ok, I found the bug in the SDK. I'm trying to solve it... I keep you informed.

**Comment by micbuffa** [comment]
Ok, this will work. I'm trying to make the syntax lighter and semantically  easier to understand using asyc/await functions. In between you can do this (if plugin1 ans plugin2 are two instances of the same plugin):
 
```
Promise.all([plugin1.load(), plugin2.load()])
    .then((nnn) => {
      // This works, the first call will load the webcomponent files
      // the second will be executed only when files are ready, so the
      // create custom element will work
      plugin1.loadGui().then((elem) => {
        console.log("Adding plugin1 GUI");
        document.body.appendChild(elem);

        // as plugin1 and plugin2 are the same, the first call to loadGUI will 
        // load the webcomponent files for its GUI, the second will not reload them
        // but as the load is async we need to embed promises...
        // I'm rewrting this using async/await to make the syntax lighter...
        plugin2.loadGui().then((elem) => {
          console.log("Adding plugin2 GUI");
          document.body.appendChild(elem);
        });
      });

      /* THIS WILL NOT WORK because the second loadGUI will try to do a createElement while the first
          call is still loading the webcomponent files...   
      Promise.all([plugin1.loadGui(), plugin2.loadGui()])
           .then((elem) => {
             console.log("Adding plugin1 GUI");
             document.body.appendChild(elem[0]);

             console.log("Adding plugin2 GUI");
             document.body.appendChild(elem[1]);
           });*/

      mediaSource.connect(nnn[0]);
      nnn[0].connect(ctx.destination)

      mediaSource.connect(nnn[1]);
      nnn[1].connect(ctx.destination);
    });

```

**Comment by micbuffa** [comment]
Ok, I found a nice fix, I'll update the SDK today. Nothing to change except that you'll be able to use Promise.all as expected, or use different sequential promises for loading multiple instances of the same plugin. I'll keep you informed. It was a nasty async problem that you found, that we encountered with the pedalboard on slow connections (some plugins would not load properly). Your finding should fix that one too as it comes from the same bug in the SDK loadGui method. Thanks a lot for contributing :-)
